Breaking: Error thrown/printed if no config found (fixes #5987)

platinumazure opened the pull request:
Just wanted to get a PR out for review. I haven't tested the CLI message yet, so I might make some tweaks over the next few days.

**Breaking change** (in case the commit message, PR title, and label aren't clear enough), so don't merge until we're ready for a major release!

mysticatea commented:
If I run the following command without .eslintrc files, how does this behave?

- `eslint . --no-eslintrc`
- `eslint . --rule semi:error`

platinumazure commented:
Goal is for the first one to not throw and run with no rules enabled (just
parse), and second one to not throw and run with semi enabled. Same with
`-c`, that should use the provided configuration. Basically if you say
anything about a configuration (including don't use eslintrc), that should
not throw- it should only throw when eslintrc is meant to be used and there
is no eslintrc anywhere (including in user's home directory).

I'll triple check those cases on an empty repository once I relearn how to
use npm link.
On Jun 12, 2016 12:00 PM, "Toru Nagashima" <notifications@github.com> wrote:

> If I run the following command without .eslintrc files, how does this
> behave?
>
>    - eslint . --no-eslintrc
>    - eslint . --rule semi:error
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/6384#issuecomment-225448030>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AARWeoxDjak-Dp9ug2e5w22hUG6RdMr2ks5qLDtDgaJpZM4IzpJl>
> .
>

mysticatea commented:
Thank you!
Sounds good to me.

platinumazure commented:
Fixed the template file with my latest commit, but now there are 18 unit test failures due to lack of config files for some tests. For any file-based ones, I will probably just throw on a `useEslintrc: false`; we'll see what else shakes out as a result.

platinumazure commented:
@IanVS Wondering if you could help me with something. Is there any reason why SourceCodeUtil might need to use configuration files that would configure a normal ESLint run on the file? For example, if the user is using babel-eslint and that is specified in a configuration file, then SourceCodeUtil would need CLIEngine to find that.

Otherwise, if we can safely assume config options are passed in by the caller, is it safe to just add `useEslintrc: false` as I have done here?

nzakas commented:
Code looks good. I think we need to make the message a lot shorter, though, no one's going to read all that. I'd do something like:

```
ESLint couldn't find a configuration file. To set up a configuration file for this project, please run:

    eslint --init
```

platinumazure commented:
Can I throw on the last line (stop by chat)? Maybe something like:

```text
ESLint couldn't find a configuration file. To set up a configuration file for this project, please run:

    eslint --init

If you think you already have a configuration file or if you need more help, please stop by the ESLint chat room: https://gitter.im/eslint/eslint
```

nzakas commented:
Fine by me.

platinumazure commented:
Apologies for the delay-- been putting my house on the market.

I hope to rebase and make the changes suggested by @nzakas this weekend.

IanVS commented:
@platinumazure, sorry for the delay, I've been a bit overwhelmed lately.

I think one possible reason why `.eslintrc` files may need to be respected is if someone has files in two directories, with one directory containing legacy script files, and another containing ES6 modules.  These directories would need to have configuration files with different `sourceType` values.  

I agree that it's probably unlikely that the parser will be changed in a config file and can be safely assumed to be provided by the caller, but I'm not so certain about `sourceType`.

What is the reason you want to add `useEslintrc: false`?

platinumazure commented:
Rebased (and it was nasty).

@IanVS I wasn't sure if .eslintrc would be needed by the config initializer. There were a few tests failing because no .eslintrc was defined for test fixtures used by the unit tests for the config initializer. I was puzzling it out and I wanted to see what happened if I just added `useEslintrc: false` to the config initializer itself. However, your reasoning has convinced me that I should instead just have the tests pass that flag to the config initializer. Thanks!

nzakas commented:
@IanVS any thoughts on @platinumazure's question?

platinumazure commented:
Please don't merge, I'm realizing that this may break config auto-create.
Hoping to confirm and fix this week.
On Jun 21, 2016 12:55 PM, "Nicholas C. Zakas" <notifications@github.com>
wrote:

> @IanVS <https://github.com/IanVS> any thoughts on @platinumazure
> <https://github.com/platinumazure>'s question?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/6384#issuecomment-227519658>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AARWei2E8O_UUwm6s4znDh7yd0mkCVpcks5qOCV7gaJpZM4IzpJl>
> .
>

IanVS commented:
Sorry, missed the question in the edit.  Seems okay to me to add an empty config, but I'll let @platinumazure test it out.

ilyavolodin commented:
@IanVS Do you mean add empty config before you start autoconfig process? What would happen if somebody does `Ctrl+C` while autoconfig is running? They will be left with an empty config. Not sure that's a good idea. However autoconfig is using Node API, and we can pass in explicit config into `verify` method. That should bypass error message for non-existing config file.

platinumazure commented:
@ilyavolodin I was originally talking about adding an .eslintrc to a
fixture folder. However, I still need to test with auto-config, at which
point your concerns become relevant.
On Jun 22, 2016 11:54 AM, "Ilya Volodin" <notifications@github.com> wrote:

> @IanVS <https://github.com/IanVS> Do you mean add empty config before you
> start autoconfig process? What would happen if somebody does Ctrl+C while
> autoconfig is running? They will be left with an empty config. Not sure
> that's a good idea. However autoconfig is using Node API, and we can pass
> in explicit config into verify method. That should bypass error message
> for non-existing config file.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/6384#issuecomment-227807120>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AARWesVKzc1xGRFdEUp_Eyu1lwvt1GYoks5qOWjTgaJpZM4IzpJl>
> .
>

nzakas commented:
We are targeting Friday for 3.0.0. Do you have an ETA? Is there something someone can help with?

nzakas commented:
If `--init` needs to be able to run without a `.eslintrc` file, then maybe we should add a `requireEslintrc` option to `CLIEngine`. We can set it to `true` when the CLI is linting normally and have it set to `false` by default so other stuff works as it does now.

platinumazure commented:
@nzakas Sorry, I've just been really busy. I won't have time tonight. I should have time tomorrow night but that's cutting it close.

And not that I really want to be holding things up, but don't you usually do release candidates before a major release? Is that not happening this time?

platinumazure commented:
@nzakas One thing I could use help on, just to know the immediate status of if this is likely to make 3.0: Can someone run `eslint --init` (and select "Inspect your JavaScript files") using the version of ESLint at the tip of the pull request branch? Basically I just want to know if we get an exception in that or not. If we do, then we might as well remove this from the 3.0.0 milestone.

ilyavolodin commented:
@platinumazure I installed eslint from your fork, and it looks like init is working. I think that due to the fact that `init` goes through Node API, it's not hitting the function you modified.

IanVS commented:
I'll catch up and look into this in an hour or so. Putting my kids to bed
now.

On Wed, Jun 22, 2016, 18:56 Ilya Volodin <notifications@github.com> wrote:

> @platinumazure <https://github.com/platinumazure> I installed eslint from
> your fork, and it looks like init is working. I think that due to the fact
> that init goes through Node API, it's not hitting the function you
> modified.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/6384#issuecomment-227902116>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AEZyAVNrZMbpVWQPZRO0H5ACuFRou-ROks5qOb2ngaJpZM4IzpJl>
> .
>

platinumazure commented:
Okay, now I'm thinking this might be safe to merge after all:

* As @ilyavolodin has noted, `--init` does not break because it goes through a different API.
* Furthermore, by adding the `.eslintrc.json` fixture file with nothing in it to the config-initializer test fixture directory, I was able to revert the change I had previously made to `source-code-util.js`, so its behavior should be unchanged.

And it helps that all tests are passing too.

So I guess my question becomes: Do we think I need more test coverage somewhere (e.g., in `--init`) to prove there is no breakage? Or do we think existing coverage is good enough? If the latter, this should be safe to merge.

@IanVS @ilyavolodin Thanks for your help. I look forward to your feedback.

IanVS commented on lib/config.js:
```diff
@@ -140,8 +141,31 @@ function getLocalConfig(thisConfig, directory) {
+        return config;
+    } else {
+
+        /**
```
Technically, I think `/**` is JSDoc notation, isn't it?  Should this perhaps just be `/*`?  Maybe it doesn't matter.

platinumazure replied:
Ack, you're right. I was misremembering the repo guidelines for non JSDoc
multiple line comments. Will fix later tonight.
On Jun 22, 2016 7:00 PM, "Ian VanSchooten" <notifications@github.com> wrote:

> In lib/config.js
> <https://github.com/eslint/eslint/pull/6384#discussion_r68156797>:
>
> > @@ -140,8 +141,31 @@ function getLocalConfig(thisConfig, directory) {
> >          config = ConfigOps.merge(localConfig, config);
> >      }
> >
> > -    // Use the personal config file if there are no other local config files found.
> > -    return found || thisConfig.useSpecificConfig ? config : ConfigOps.merge(config, getPersonalConfig());
> > +    if (found || thisConfig.useSpecificConfig) {
> > +        return config;
> > +    } else {
> > +
> > +        /**
>
> Technically, I think /** is JSDoc notation, isn't it? Should this perhaps
> just be /*? Maybe it doesn't matter.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/6384/files/94bdc2000dfe96c13fe9b8cb9260364753f2dfff#r68156797>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AARWegqLLQzvcBtMqekqxJdSD_K42Dzjks5qOcyPgaJpZM4IzpJl>
> .
>

nzakas replied:
Yeah, can you update that to be non-JSDoc?

platinumazure replied:
@nzakas :+1:

IanVS commented on messages/no-config-found.txt:
```diff
@@ -0,0 +1,21 @@
+We couldn't find a valid configuration file for this linting run.
+
+This is the directory we were trying to lint: <%= directory %>
```
It might be slightly more accurate to say something like `We looked for configuration files in <%= directory =%> and its ancestors.`

platinumazure replied:
@IanVS :+1:

IanVS commented on tests/fixtures/source-code-util/.eslintrc.json:
```diff
@@ -0,0 +1,3 @@
+{
+    // This file is only to ensure no error is thrown due to lack of config.
```
I really like that you've included a note explaining why this file exists, but technically comments are not valid in `.json` files (hence the red highlighting from GitHub).  Perhaps it would be better to make this a `.yaml` or `.js` file?

nzakas replied:
We allow comments in JSON config files, we just strip them out before parsing

platinumazure replied:
Strong preference to switch to "__comment"-style?

IanVS replied:
Not sure what you mean, but I think there's no need to change this.

platinumazure replied:
@IanVS Acknowledged, won't change.

What I was referring to is, since JSON doesn't support comments, some folks will just add a top-level key "__comment" (or similar) to store their file comments.

nzakas commented:
@platinumazure yeah, we have typically done preview releases, but they are a huge pain in the ass because there is a lot of manual process to that. For this one we are trying to do a much smaller release with fewer breaking changes to try to avoid that mess.

nzakas commented:
@platinumazure also, the message needs to be updated per https://github.com/eslint/eslint/pull/6384#issuecomment-225640214

The code LGTM.

IanVS commented:
@ilyavolodin I was not suggesting to add an empty config file as part of `--init`, but rather like @platinumazure said, to add it to the test fixture folder to prevent unrelated tests from failing.  

@platinumazure There are a lot of tests, so I might have missed one that already exists, but I think it would be good to have a few tests to verify the cases that @mysticatea asked about in https://github.com/eslint/eslint/pull/6384#issuecomment-225448030.  This would mean taking the test you made which verifies an error is thrown, and adding `useEslintrc: false` to verify that an error is not thrown, (and the same for specifying a rule config).  Other than that, I think we have pretty good test coverage and can be reasonably assured we won't have other breakages.

platinumazure commented:
@IanVS @nzakas Fixed the JSDoc issue and changed the message. I decided to use @nzakas' message but still mention the directory being examined. @nzakas Please let me know if you think it's now too long as a result.

This is probably all I can do from work. I'll look at the tests tonight. I think the only thing I'm really missing is an explicit `useEslintrc: false` test-- there should be tests already for actually resolving configurations, in current/ancestor/home directories, which should continue to work.

nzakas commented:
@platinumazure is it just the one test that's missing? If so, can you describe it a bit more? Maybe someone can take a stab at it.

IanVS commented:
It's a test to make sure there is no error if someone runs `eslint --no-eslintrc` in a directory without a config file.  I'll take a shot at it real quick.

nzakas commented:
Cool. I'd just like to get this merged today so we can do some verification on master before releasing tomorrow. :+1:

IanVS commented:
I opened another PR, because I couldn't figure out a way to add to this one, since it was from @platinumazure's repo. 

https://github.com/eslint/eslint/pull/6520

Make sure @platinumazure gets the credit for the work though.
